includes/class-wc-stripe-payment-tokens.php

```diff
@@ -262,28 +267,33 @@ public function woocommerce_get_customer_upe_payment_tokens( $tokens, $user_id,
-
-			// Prevent unnecessary recursion, WC_Payment_Token::save() ends up calling 'woocommerce_get_customer_payment_tokens' in some cases.
-			remove_action( 'woocommerce_get_customer_payment_tokens', [ $this, 'woocommerce_get_customer_payment_tokens' ], 10, 3 );
-			foreach ( $payment_methods as $payment_method ) {
-				if ( ! isset( $remaining_tokens[ $payment_method->id ] ) ) {
-					$payment_method_type = $this->get_original_payment_method_type( $payment_method );
-					if ( ! in_array( $payment_method_type, $reusable_payment_methods, true ) ) {
-						continue;
+		try {
+			foreach ( $retrievable_payment_method_types as $payment_method_id ) {
+				$payment_methods = $customer->get_payment_methods( $payment_method_id );
+
+				// Prevent unnecessary recursion, WC_Payment_Token::save() ends up calling 'woocommerce_get_customer_payment_tokens' in some cases.
+				remove_action( 'woocommerce_get_customer_payment_tokens', [ $this, 'woocommerce_get_customer_payment_tokens' ], 10, 3 );
+				foreach ( $payment_methods as $payment_method ) {
+					if ( ! isset( $remaining_tokens[ $payment_method->id ] ) ) {
+						$payment_method_type = $this->get_original_payment_method_type( $payment_method );
+						if ( ! in_array( $payment_method_type, $reusable_payment_methods, true ) ) {
+							continue;
+						}
+						// Create new token for new payment method and add to list.
+						$upe_payment_method         = $gateway->payment_methods[ $payment_method_type ];
+						$token                      = $upe_payment_method->create_payment_token_for_user( $user_id, $payment_method );
+						$tokens[ $token->get_id() ] = $token;
+					} else {
+						// Count that existing token for payment method is still present on Stripe.
+						// Remaining IDs in $remaining_tokens no longer exist with Stripe and will be eliminated.
+						unset( $remaining_tokens[ $payment_method->id ] );
 					}
-					// Create new token for new payment method and add to list.
-					$upe_payment_method         = $gateway->payment_methods[ $payment_method_type ];
-					$token                      = $upe_payment_method->create_payment_token_for_user( $user_id, $payment_method );
-					$tokens[ $token->get_id() ] = $token;
-				} else {
-					// Count that existing token for payment method is still present on Stripe.
-					// Remaining IDs in $remaining_tokens no longer exist with Stripe and will be eliminated.
-					unset( $remaining_tokens[ $payment_method->id ] );
 				}
+				add_action( 'woocommerce_get_customer_payment_tokens', [ $this, 'woocommerce_get_customer_payment_tokens' ], 10, 3 );
 			}
-			add_action( 'woocommerce_get_customer_payment_tokens', [ $this, 'woocommerce_get_customer_payment_tokens' ], 10, 3 );
+		} catch ( WC_Stripe_Exception $e ) {
+			wc_add_notice( $e->getLocalizedMessage(), 'error' );
+			WC_Stripe_Logger::log( 'Error: ' . $e->getMessage() );
 		}
 
 		// Eliminate remaining payment methods no longer known by Stripe.
```

Comment:
Should we add these lines (https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2668/files#diff-f80280bf89ae5350f2729f8c88d0d39cd4df34c7f13fe87db1445bcb03d6ac34R299-R306) that eliminate the remaining payment methods within the `try` block?

These lines would still execute if an exception is thrown. We unset some items within $remaining_tokens down  in the `try` block (https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2668/files#diff-f80280bf89ae5350f2729f8c88d0d39cd4df34c7f13fe87db1445bcb03d6ac34R289). My concern is that, if an exception is thrown, some items of $remaining_tokens won't be unset and may get unexpectedly deleted.

Reply:
Agree with this. Moved this part within the try/catch. :heavy_check_mark: